Audit Fixed QS-14

arcantheon opened the pull request:
Fixed repititive inheritence.

![Screenshot 2024-06-20 at 20 37 55](https://github.com/Sperax/Demeter-Protocol/assets/140178288/d09526bd-7695-4363-8795-63bf8e823f38)

arcantheon committed:
perf(Farms): removed repetitive inheritence

arcantheon committed:
Merge branch 'dev' into perf/inheritence

arcantheon committed:
fix(state of implementation contracts)

arcantheon committed:
Merge branch 'dev' into perf/inheritence

YashP16 reviewed (changes requested):
@arcantheon,
Let's remove ExpirableFarm from the OperableDeposit and resolve these dependencies at the topmost level only.

parv3213 committed:
fix: make `_updateFarmRewardData` public and rename to `updateFarmRewardData` (#74)

1. `updateFarmRewardData` will allow users/dapp to call this function externally to accrue rewards.
2. `updateFarmRewardData` function can now also be triggered externally before `farmEndTime` to accrue user rewards. Which otherwise cannot be accrued after farm expiry.

Co-authored-by: arcantheon <140178288+arcantheon@users.noreply.github.com>

parv3213 committed:
feat: remove LegacyFarmRewarder (#81)

arcantheon committed:
atomic inheritence, operable deposit and expirable farms

arcantheon committed:
Merge branch 'dev' into perf/inheritence

arcantheon commented:
> @arcantheon, Let's remove ExpirableFarm from the OperableDeposit and resolve these dependencies at the topmost level only.

Done: e8fa68ca33fa89234f8b9a315f3532a402616f62

parv3213 commented on contracts/e20-farms/E20Farm.sol:
```diff
@@ -114,10 +92,28 @@ abstract contract E20Farm is ExpirableFarm, OperableDeposit {
     /// @notice Recover erc20 tokens other than the reward Tokens and farm token.
     /// @param _token Address of token to be recovered.
-    function recoverERC20(address _token) external override onlyOwner nonReentrant {
+    function recoverERC20(address _token) public virtual override onlyOwner nonReentrant {
```
If instead of overriding this we override `_recoverE20` we can get way without making this function public

arcantheon replied:
Good point, done: da7d7f2914e0beec9037c71e1f95ebe8e2f39aee

parv3213 commented on contracts/rewarder/Rewarder.sol:
```diff
@@ -96,7 +96,9 @@ contract Rewarder is Ownable, Initializable, ReentrancyGuard {
     error CalibrationRestricted(address farm);
 
-    constructor() Ownable(msg.sender) {}
+    constructor() Ownable(msg.sender) {
```
Let's handle this change in QS-10?

parv3213 replied:
https://github.com/Sperax/Demeter-Protocol/pull/72#discussion_r1652667664

arcantheon replied:
If we merge QS10 first, this won't be reflecting as a change anymore.

arcantheon replied:
Reverting pushed changes would waste time and can introduce bugs.

parv3213 commented on contracts/Farm.sol:
Actually even this change should be part of QS-10. Btw, it is already there in QS-10

parv3213 replied:
@arcantheon I have pushed a commit to remove this and related code in [cb6fb02](https://github.com/Sperax/Demeter-Protocol/pull/72/commits/cb6fb027e25a299542618ee31f3549f850b114c4). Please check and if you feel otherwise we can revert this commit. 
Btw, this commit is, although passing all test, failing a solhint check. I haven't fixed this as we might change some logic based on https://github.com/Sperax/Demeter-Protocol/pull/72#discussion_r1652658741

arcantheon replied:
Makes sense

parv3213 commented on contracts/e20-farms/E20Farm.sol:
```diff
@@ -114,10 +92,28 @@ abstract contract E20Farm is ExpirableFarm, OperableDeposit {
+        uint256 _cooldownPeriod,
+        address _farmToken,
+        RewardTokenData[] memory _rwdTokenData
+    ) public {
```
Even this function we are making `public` from `external`

parv3213 replied:
Here also can we make `initialize` in E20Farm an internal function? And call this internal function in top layer farms? Also, overriding does not makes sense here both functions serves a different purpose.

arcantheon replied:
Done: c9acb2278965b61a42fe9360652743b8317cd48b

parv3213 committed:
revert: reverting all changes related to changing initializer modifier

- This change is better suited for another PR

arcantheon committed:
refactor(recoverERC20): internal function instead of external function

arcantheon committed:
Moved onlyOwner modifier to the lower level function call

parv3213 commented on contracts/Farm.sol:
```diff
@@ -170,7 +170,7 @@ abstract contract Farm is FarmStorage, Ownable, ReentrancyGuard, Initializable,
     /// @notice Recover erc20 tokens other than the reward Tokens.
     /// @param _token Address of token to be recovered.
     function recoverERC20(address _token) external virtual onlyOwner nonReentrant {
```
```suggestion
    function recoverERC20(address _token) external onlyOwner nonReentrant {
```

arcantheon committed:
refactor(E20 farms): initialization

arcantheon committed:
Update contracts/Farm.sol

Co-authored-by: Parv Garg <parv3213@gmail.com>

YashP16 reviewed (approved):
Changes look good

parv3213 reviewed (approved):
Looks good 👍 .
There are two solhint warnings introduced with these changes. Can you please fix them as well.
